Assets/Project/Scripts/Common/Utils/StageStatusService.cs

```diff
@@ -0,0 +1,91 @@
+using System.Collections.Generic;
+using System.Linq;
+using Cysharp.Threading.Tasks;
+using Treevel.Common.Entities;
+using Treevel.Common.Entities.GameDatas;
+using Treevel.Common.Managers;
+using Treevel.Common.Networks;
+using Treevel.Common.Networks.Requests;
+using UnityEngine;
+
+namespace Treevel.Common.Utils
+{
+    public sealed class StageStatusService
+    {
+        // For Singleton
+        private StageStatusService() { }
+
+        /// <summary>
+        /// インスタンス
+        /// </summary>
+        public static readonly StageStatusService Instance = new StageStatusService();
+
+        /// <summary>
+        /// オンメモリに StageStatus を保持する
+        /// </summary>
+        private readonly Dictionary<string, StageStatus> _cachedStageStatusDic = new Dictionary<string, StageStatus>();
+
+        /// <summary>
+        /// 全ステージの StageStatus を読み込んでおく
+        /// </summary>
+        public async UniTask PreloadAllStageStatusesAsync()
+        {
+            var stageStatuses = await NetworkService.Execute(new GetAllStageStatusListRequest());
+            var stageStatusList = stageStatuses as List<StageStatus> ?? stageStatuses.ToList();
+
+            foreach (var stageData in GameDataManager.GetAllStages()) {
+                var stageStatus = stageStatusList.FirstOrDefault(stageStatus => stageStatus.treeId == stageData.TreeId
+                                                                                && stageStatus.stageNumber == stageData.StageNumber)
+                                  ?? new StageStatus(stageData.TreeId, stageData.StageNumber);
+
+                _cachedStageStatusDic[StageData.EncodeStageIdKey(stageData.TreeId, stageData.StageNumber)] = stageStatus;
+            }
+        }
+
+        /// <summary>
+        /// 特定の StageStatus を取得する
+        /// </summary>
+        /// <param name="treeId"> 木の Id </param>
+        /// <param name="stageNumber"> ステージ番号 </param>
+        /// <returns> StageStatus </returns>
+        public StageStatus Get(ETreeId treeId, int stageNumber)
+        {
+            return _cachedStageStatusDic.Values.ToList()
+                    .First(stageStatus => stageStatus.treeId == treeId && stageStatus.stageNumber == stageNumber);
```

Comment:
`imo` どっちが性能良いか分からないけど、Dictionaryならこれで良いのでは、、、と思ってしまう。
```suggestion
            return _cachedStageStatusDic[StageData.EncodeStageIdKey(treeId, stageNumber)];
```

Reply:
cc76b38